xdsl/backend/riscv/prologue_epilogue_insertion.py

```diff
@@ -0,0 +1,92 @@
+from xdsl.builder import Builder
+from xdsl.dialects import builtin, riscv, riscv_func
+from xdsl.dialects.riscv import (
+    FloatRegisterType,
+    IntRegisterType,
+    Registers,
+    RISCVRegisterType,
+)
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+
+
+class PrologueEpilogueInsertion(ModulePass):
+    """
+    Pass inserting a prologue and epilogue according to the RISC-V ABI.
+    The prologues and epilogues are responsible for saving any callee-preserved
+    registers.
+    In RISC-V these are 's0' to 's11' and 'fs0' to `fs11'.
+    The stack pointer 'sp' must also be restored to its original value.
+
+    This pass should be run late in the pipeline after register allocation.
+    It does not itself require register allocation nor invalidate the result of the
+    register allocator.
+    """
+
+    name = "riscv-prologue-epilogue-insertion"
+
+    def _process_function(self, func: riscv_func.FuncOp) -> None:
+        # Find all callee-preserved registers that are clobbered. We define clobbered
+        # as it being the result of some operation and therefore written to.
+        used_callee_preserved_registers = dict[
+            IntRegisterType | FloatRegisterType, None
+        ]()
+        for op in func.walk():
+            # Special cases ops that do not actually write to the register.
+            if isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp):
+                continue
+
+            for res in op.results:
+                if res.type in Registers.S or res.type in Registers.FS:
+                    used_callee_preserved_registers[res.type] = None
```

Comment:
```suggestion
        used_callee_preserved_registers = {
            res.type
            for op in func.walk()
            if not isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp)
            for res in op.results
            if res.type in Registers.S or res.type in Registers.FS
        }
```
Might as well use a set here instead of a dict to None, no?

Reply:
This would make the compiler output not-deterministic as sets have unspecified iteratior order

Reply:
Then you'd need to make sure to sort it before you use it otherwise you get funky stuff :ghost:

Reply:
Sorted did not work immediately as there is no comparison operator between `IntRegisterType` and `FloatRegisterType`. I applied your changes but kept it as a dict for now. I can add that comparison operator (or can this be auto generated by some annotator?) if you feel strongly about this